aspects/transaction.go

```diff
@@ -0,0 +1,131 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"sync"
+)
+
+type DatabagRead func() (JSONDataBag, error)
+type DatabagWrite func(JSONDataBag) error
+
+// Transaction performs read and writes to a databag in an atomic way.
+type Transaction struct {
+	pristine JSONDataBag
+	deltas   []map[string]interface{}
+	schema   Schema
+
+	readDatabag  DatabagRead
+	writeDatabag DatabagWrite
+	mu           sync.RWMutex
+}
+
+// NewTransaction takes a getter and setter to read and write the databag.
+func NewTransaction(readDatabag DatabagRead, writeDatabag DatabagWrite, schema Schema) (*Transaction, error) {
+	databag, err := readDatabag()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Transaction{
+		pristine:     databag.Copy(),
+		schema:       schema,
+		readDatabag:  readDatabag,
+		writeDatabag: writeDatabag,
+	}, nil
+}
+
+// Set sets a value in the transaction's databag. The change isn't persisted
+// until Commit returns without errors.
+func (t *Transaction) Set(path string, value interface{}) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.deltas = append(t.deltas, map[string]interface{}{path: value})
+	return nil
+}
+
+// Get reads a value from the transaction's databag including uncommitted changes.
+func (t *Transaction) Get(path string, value interface{}) error {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	copy := t.pristine.Copy()
+	if err := applyDeltas(copy, t.deltas); err != nil {
+		return err
+	}
+
+	return copy.Get(path, value)
```

Comment:
Uncommitted changes are not schema validated. Is that intentional? Not sure how this is planned to be used.

Reply:
They are validated by the JSONDataBag when they're written and then the whole thing is validated on Commit